### PR TITLE
Add a setting to toggle flashing on pages with images

### PR DESCRIPTION
**Pull request by sargunv**
See https://github.com/koreader/koreader/issues/10047 for issue / discussion.

To figure out how to add a setting, I just copied another existing setting (always flash on chapter boundaries).

Tested by adding a log line to the code that flashes. Unsure how to get this dev build onto my Android e-ink device to actually test the flashing / not flashing.

~~Also unsure how to make this setting default to "on", but if we want to preserve existing behavior by default, that should be the default.~~ Fixed.

~~Also unsure about i10n files: do I run some script to add this entry to every language file? Do I add the entries manually? What do I fill for languages I can't translate to?~~

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/10049)

### CODE REVIEW

**Comment by NiLuJe** [comment]
> Also unsure how to make this setting default to "on", but if we want to preserve existing behavior by default, that should be the default.

`isTrue` -> `nilOrTrue`


> about i10n

There's nothing to be done, it's all automated.

**Comment by sargunv** [comment]
> `isTrue` -> `nilOrTrue`

I tried this, but it broke toggling the setting. Perhaps I need to implement `onToggleFlashOnImages` and `onSetFlashOnImages` differently?

**Comment by sargunv** [comment]
Yup, that was it.

**Comment by Frenzie** [comment]
I believe we use the word refresh, not flash.

The translation is extracted automatically when the PR is merged.

**Comment by sargunv** [comment]
I used the same naming conventions as the neighboring setting "Always flash on chapter boundaries". That's "refresh" in the setting key but "flash" in the displayed text and event/function name. Should I still change this to use the word "refresh" even though that'd be inconsistent with the others?

**Comment by NiLuJe** [comment]
IIRC, only the "refresh rate" setting uses refresh to mean flash (which I hate with a fiery passion, but it's old ;p).

TL;DR: I *vastly* prefer flash for user-visible stuff.

**Comment by sargunv** [comment]
👍 will keep it as "flash" then (I have no opinion here other than preferring consistency)

File: frontend/dispatcher.lua [hunk 1]
```diff
@@ -274,6 +276,8 @@ local dispatcher_menu_order = {
     "toggle_flash_on_chapter_boundaries",
     "set_no_flash_on_second_chapter_page",
     "toggle_no_flash_on_second_chapter_page",
+    "set_flash_on_images",
+    "toggle_flash_on_images",
```
Comment (poire-z) [inline frontend/dispatcher.lua:280]: Here for the actions `...flash_on_images` but also for the setting name `refresh_on_images`: may be have them more verbose and explicite and long, like the others: 
```
refresh_on_chapter_boundaries      / ...flash_on_chapter_boundaries
no_refresh_on_second_chapter_page  / ...no_flash_on_second_chapter_page
refresh_on_page_with_images        / ...flash_on_page_with_images
```
so when browsing around settings and gesture files, we know they are related to pages of books, and not just a refresh on icons or images collections books like CBZ.

Comment (sargunv) [reply frontend/dispatcher.lua]: Done!

**Comment by mergen3107** [comment]
@sargunv 
This is actually great PR!

Thank you! Will be happy to test it out. 

I have a config with fb2mobi converter which can make epubs as well. This config also adds nice custom vignettes to the chapter titles and at the end of chapters. So each time such vignettes appear on page, even on mid-chapter titles of h3, the screen flashes. 

Now with this setting this can be avoided :) Onyx Boox Kon-Tiki 2 that I have has a very good long term eink performance anyway, so currently all my refreshes are kept to minimum in KOReader.

**Comment by sargunv** [comment]
Glad it's helpful! 

My motivation is similar; it came from reading a book with chapter icons like these on a Boox Leaf 2:

Screenshot_20230125-183942

File: frontend/device/devicelistener.lua [hunk 2]
```diff
@@ -306,6 +306,18 @@ function DeviceListener:onToggleNoFlashOnSecondChapterPage()
...
+function DeviceListener:onSetFlashOnPagesWithImages(toggle)
+    if toggle == true then
+        _unsetSetting("refresh_on_pages_with_images")
+    else
+        G_reader_settings:makeFalse("refresh_on_pages_with_images")
+    end
+end
+
+function DeviceListener:onToggleFlashOnPagesWithImages()
+    G_reader_settings:flipNilOrTrue("refresh_on_pages_with_images")
```
Comment (poire-z) [inline frontend/device/devicelistener.lua:318]: This looks unbalanced, some of your branches using G_reader_settings while others use _unsetSettings - while all other non-yours branches use the wrappers _unsetSetting et al.
Indeed, none of these wrappers provide wrappers for flipNilOrTrue or makeFalse - and they look a bit useless doing nothing else than G_readersetting:make/flip... that one could do directly at the call site.
So, it's for @NiLuJe , who added them, to say how to proceeed :) (my preference would be to get rid of these wrappers).

Comment (sargunv) [reply frontend/device/devicelistener.lua:318]: I think refactoring the wrappers to encapsulate the repetitive if/then on `onSet*` functions and support both cases (default=false, default=true) might make sense.

Comment (sargunv) [reply frontend/device/devicelistener.lua]: I've pushed a commit to do so. Happy to revert if @NiLuJe prefers to keep them.

Comment (NiLuJe) [reply frontend/device/devicelistener.lua]: Looks fine to me ;).

(I don't recall the full history of those weird wrappers, but they might have been doing more stuff at one point in time, and just became on-liners when I tweaked the *Settings API...)

**Review by NiLuJe — APPROVED** [review]

File: frontend/device/devicelistener.lua (hunk 2 above)
Comment (NiLuJe) [reply frontend/device/devicelistener.lua]: > (I don't recall the full history of those weird wrappers, but they might have been doing more stuff at one point in time, and just became one-liners when I tweaked the *Settings API...)

Or it was simply someone allergic to the fairly heinous naming scheme of said APIs ;p.

**Review by NiLuJe — APPROVED** [review]
> Reviewed 3 of 4 files at r3, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/10049)* status: all files reviewed, 2 unresolved discussions (waiting on @poire-z)
